scrapy/utils/gz.py

```diff
@@ -9,32 +9,58 @@
 import six
 
 
-# - Python>=3.5 GzipFile's read() has issues returning leftover
-#   uncompressed data when input is corrupted
-#   (regression or bug-fix compared to Python 3.4)
-# - read1(), which fetches data before raising EOFError on next call
-#   works here but is only available from Python>=3.3
-# - scrapy does not support Python 3.2
-# - Python 2.7 GzipFile works fine with standard read() + extrabuf
-if six.PY2:
-    def read1(gzf, size=-1):
-        return gzf.read(size)
-else:
-    def read1(gzf, size=-1):
-        return gzf.read1(size)
+class GzipFilePartial(GzipFile):
+    """A subclass that patches GzipFile to be able to handle broken/partial
+    archives by replacing GzipFile._read_eof with a no-op. 
+    
+    Python3: read1() uses the method GzipFile._check_closed(), so here we replace
+    that with a no-op.
+    This is done so that GzipFile doesn't do it's checksum comparison on broken
+    archives, which would result in the decompression failing.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(GzipFilePartial, self).__init__(*args, **kwargs)
+
+    # - Python>=3.5 GzipFile's read() has issues returning leftover
+    #   uncompressed data when input is corrupted
+    #   (regression or bug-fix compared to Python 3.4)
+    # - read1(), which fetches data before raising EOFError on next call
+    #   works here but is only available from Python>=3.3
+    # - scrapy does not support Python 3.2
+    # - Python 2.7 GzipFile works fine with standard read() + extrabuf
+    def readHandler(self, size=-1):
+        try:
+            if six.PY2:
+                return self.read(size)
+            else:
+                return self.read1(size)
+        except(IOError):
+            # replace the _read_eof method with a no-op and try the read again,
+            # would make a partial archive read work
+            try:
+                if six.PY2:
+                    self._read_eof = lambda *args, **kwargs: None
```

Comment:
@kmike I like this approach of patching the class method here, instead of overwriting it per se. It is not gonna be used again (on this instance) anyways. You think it's okay?